### PR TITLE
Improve composer visiblity

### PR DESCRIPTION
**Pull request by gsouquet**
Fixes vector-im/element-web#22072
Fixes https://github.com/vector-im/element-web/issues/17362

Screen Shot 2022-05-13 at 11 12 37
Screen Shot 2022-05-13 at 11 12 52
Screen Shot 2022-05-13 at 11 13 01

@niquewoodhouse I've implemented an enter transition for the send button, let me know whether you want to tweak the different values I've added. I believe I landed somewhere quite nice

---
Here's what your changelog entry will look like:

## ✨ Features
 * Improve composer visiblity ([\#8578](https://github.com/matrix-org/matrix-react-sdk/pull/8578)). Fixes vector-im/element-web#22072 and vector-im/element-web#17362.

### CODE REVIEW

File: res/css/structures/_RightPanel.scss [hunk 1]
```diff
@@ -30,6 +30,11 @@ limitations under the License.
     .mx_RoomView_MessageList {
         padding: 14px 18px; // top and bottom is 4px smaller to balance with the padding set above
     }
+
+    .mx_MessageComposer_wrapper {
+        margin-top: 8px;
+        margin-bottom: 8px;
```
Comment (t3chguy) [inline res/css/structures/_RightPanel.scss:36]:
```suggestion
        margin-top: $spacing-8;
        margin-bottom: $spacing-8;
```

File: res/css/structures/_RightPanel.scss [hunk 2]
```diff
@@ -177,14 +182,14 @@ $pulse-color: $alert;
 
 .mx_RightPanel_headerButton_unread {
     &::before {
-        background-color: $room-icon-unread-color !important;
+        background-color: $room-icon-unread-color  !important;
```
Comment (t3chguy) [inline res/css/structures/_RightPanel.scss:185]: why double-space?

File: res/css/structures/_RightPanel.scss [hunk 3]
```diff
@@ -177,14 +182,14 @@ $pulse-color: $alert;
...
     }
 }
 
 .mx_RightPanel_headerButton_highlight,
 .mx_RightPanel_headerButton:hover {
     &::before {
-        background-color: $accent !important;
+        background-color: $accent  !important;
```
Comment (t3chguy) [inline res/css/structures/_RightPanel.scss:192]: why double-space?

File: res/css/structures/_RoomView.scss [hunk 4]
```diff
@@ -30,6 +30,13 @@ limitations under the License.
     flex-direction: column;
     flex: 1;
     position: relative;
+
+}
+
+.mx_MainSplit_timeline {
+    .mx_MessageComposer_wrapper {
+        margin: 16px;
```
Comment (t3chguy) [inline res/css/structures/_RoomView.scss:38]:
```suggestion
        margin: $spacing-16;
```

File: res/css/structures/_RoomView.scss [hunk 5]
```diff
@@ -155,7 +162,7 @@ limitations under the License.
     .mx_RoomView_messageListWrapper {
         justify-content: flex-start;
 
-        >.mx_RoomView_MessageList > li > ol {
+        >.mx_RoomView_MessageList>li>ol {
```
Comment (t3chguy) [inline res/css/structures/_RoomView.scss:165]: why removing these spaces?

File: res/css/structures/_RoomView.scss [hunk 6]
```diff
@@ -266,7 +273,7 @@ hr.mx_RoomView_myReadMarker {
 }
 
 .mx_RoomView_ongoingConfCallNotification a {
-    color: $accent-fg-color !important;
+    color: $accent-fg-color  !important;
```
Comment (t3chguy) [inline res/css/structures/_RoomView.scss:276]: why double-space?

File: res/css/views/rooms/_MessageComposer.scss [hunk 7]
```diff
@@ -20,8 +20,30 @@ limitations under the License.
...
+        flex: 1;
+    }
+
+    >[role=button] {
+        margin-bottom: 7px;
+    }
+
+    .mx_VoiceRecordComposerTile_delete {
+        margin-bottom: 9px;
+    }
```
Comment (t3chguy) [inline res/css/views/rooms/_MessageComposer.scss:37]: can we avoid spacings which don't fit into our $spacing-vars?

File: res/css/views/rooms/_MessageComposer.scss [hunk 8]
```diff
@@ -20,8 +20,30 @@ limitations under the License.
...
+
+    .mx_VoiceRecordComposerTile_stop,
+    .mx_MessageComposer_sendMessage {
+        margin-bottom: 4px;
```
Comment (t3chguy) [inline res/css/views/rooms/_MessageComposer.scss:41]:
```suggestion
        margin-bottom: $spacing-4;
```

File: res/css/views/rooms/_MessageComposer.scss [hunk 9]
```diff
@@ -20,8 +20,30 @@ limitations under the License.
...
+    }
+
+    .mx_VoiceMessagePrimaryContainer {
+        margin-right: 8px;
```
Comment (t3chguy) [inline res/css/views/rooms/_MessageComposer.scss:45]:
```suggestion
        margin-right: $spacing-8;
```

File: res/css/views/rooms/_MessageComposer.scss [hunk 10]
```diff
@@ -56,7 +78,16 @@ limitations under the License.
     display: flex;
     flex-direction: row;
     align-items: center;
-    width: 100%;
+    border: 1px solid $quaternary-content;
+    border-radius: 8px;
+    padding: 12px 8px;
```
Comment (t3chguy) [inline res/css/views/rooms/_MessageComposer.scss:83]:
```suggestion
    padding: $spacing-12 $spacing-8;
```

File: res/css/views/rooms/_MessageComposer.scss [hunk 11]
```diff
@@ -56,7 +78,16 @@ limitations under the License.
...
+    margin-right: 16px;
```
Comment (t3chguy) [inline res/css/views/rooms/_MessageComposer.scss:84]:
```suggestion
    margin-right: $spacing-16;
```

File: res/css/views/rooms/_MessageComposer.scss [hunk 12]
```diff
@@ -121,16 +150,22 @@ limitations under the License.
 
 // FIXME: rather unpleasant hack to get rid of <p/> margins.
 // really we should be mixing in markdown-body from gfm.css instead
-.mx_MessageComposer_editor > :first-child {
+.mx_MessageComposer_editor> :first-child {
```
Comment (t3chguy) [inline res/css/views/rooms/_MessageComposer.scss:153]: Why removing the space?

File: res/css/views/rooms/_MessageComposer.scss [hunk 13]
```diff
@@ -121,16 +150,22 @@ limitations under the License.
...
     margin-top: 0 !important;
 }
-.mx_MessageComposer_editor > :last-child {
+
+.mx_MessageComposer_editor> :last-child {
```
Comment (t3chguy) [inline res/css/views/rooms/_MessageComposer.scss:157]: Why removing the space?

File: res/css/views/rooms/_MessageComposer.scss [hunk 14]
```diff
@@ -166,15 +201,17 @@ limitations under the License.
...
     color: $accent;
 }
 
 .mx_MessageComposer_button_highlight {
     background: rgba($accent, 0.25);
+
     // make the icon the accent color too
     &::before {
-        background-color: $accent !important;
+        background-color: $accent  !important;
```
Comment (t3chguy) [inline res/css/views/rooms/_MessageComposer.scss:214]: Why double-space?

**Review by niquewoodhouse — CHANGES_REQUESTED** [review]
Some small changes requested, but need to think on a couple of other points.  I'm in Safari if it matters.

- [ ] Vertical alignment of icons to right of input 

image

- [ ] Appearing of send button feels inconsistent

So at the moment it feels like it appears when input is active.  When I arrive at a room, its visible, because the input is on focus.  But when I send the message, the input is still on focus but the send button has gone.  Can we easily try it with just appearing when there's content in the input?

- [x] Transition the exit of the send button

Its currently transitioning in but its not transitioning out.  I think that would be a smoother experience.

File: res/css/structures/_RightPanel.scss (hunk 2 above)
Comment (gsouquet) [reply res/css/structures/_RightPanel.scss:185]: I had the VSCode autoformatter for SCSS enabled. That messed up a couple of stuff. reverting all of that

**Review by t3chguy — APPROVED** [review]

**Comment by gsouquet** [comment]
Wasn't able to reproduce the button alignment issue on Safari 15.4

**Comment by niquewoodhouse** [comment]
> Wasn't able to reproduce the button alignment issue on Safari 15.4

I also can't reproduce this issue in 15.5 so will cross out

**Comment by kittykat** [comment]
@niquewoodhouse I think it's best to raise it with the design team as there should be a designer on every development team so they should be aware of anyone else with an interest in this area

**Review by niquewoodhouse — COMMENTED** [review]
Just recording some issues for us to solve.

- [ ] Reply behaviour 

Reply field seems to get squashed up, depending on the length of the message being replied to.

image

image

(Added a design to Figma file)

image

**Comment by gsouquet** [comment]
@niquewoodhouse here's a lil screenshot of the new look and feel. The display name is using the `SenderProfile` component, and uses the same color code as it normally do

Screen Shot 2022-06-01 at 11 08 37

File: src/components/views/elements/AccessibleButton.tsx [hunk 15]
```diff
@@ -41,7 +41,7 @@ type AccessibleButtonKind = | 'primary'
  * onClick:  (required) Event handler for button activation. Should be
  *           implemented exactly like a normal onClick handler.
  */
-interface IProps extends React.InputHTMLAttributes<Element> {
+export interface IProps extends React.InputHTMLAttributes<Element> {
```
Comment (t3chguy) [inline src/components/views/elements/AccessibleButton.tsx:44]: We shouldn't be exporting this, please use `React.ComponentProps`

File: src/components/views/messages/DisambiguatedProfile.tsx [hunk 16]
```diff
@@ -57,13 +62,14 @@ export default class DisambiguatedProfile extends React.Component<IProps> {
...
-        );
+        return React.createElement(this.props.as, {
+            className: "mx_DisambiguatedProfile",
+            onClick,
+        }, [
+            <span className={displayNameClasses} dir="auto" key="displayNameClasses">
+                { rawDisplayName }
+            </span>,
+            mxidElement,
+        ]);
```
Comment (t3chguy) [inline src/components/views/messages/DisambiguatedProfile.tsx:73]: Won't need these keys if using an array? Can't we use a fragment?

File: src/components/views/rooms/ReplyPreview.tsx [hunk 17]
```diff
@@ -44,19 +46,17 @@ export default class ReplyPreview extends React.Component<IProps> {
...
-                        onClick={() => cancelQuoting(this.context.timelineRenderingType)}
-                    />
-                </div>
-                <ReplyTile
-                    mxEvent={this.props.replyToEvent}
-                    permalinkCreator={this.props.permalinkCreator}
-                />
+            <div className="mx_ReplyPreview_header">
+                <ReplyIcon />
+                { _t('Reply to') } &nbsp;
```
Comment (t3chguy) [inline src/components/views/rooms/ReplyPreview.tsx:51]: This needs to include the sender profile as a `<tag/>` as the order of the sentence will vary per-language

**Review by t3chguy — CHANGES_REQUESTED** [review]
See line-comments

Opinion time:
image
It is really easy to miss the border of this, previously there was a drop shadow which helped separate it from the timeline
image

Vertical alignment feels odd
image

**Review by t3chguy — APPROVED** [review]

**Review by niquewoodhouse — CHANGES_REQUESTED** [review]
Took the time to do a more thorough review:

### When lacking permission to post

- [ ] Default state

composer_alt-01-no-permission-1@2x

- [ ] Hover state (cursor change)

composer_alt-01-no-permission-1-hover@2x

**When room "upgraded"**

composer-room-replaced@2x

**Reply mode**

The earlier feedback from @t3chguy on the lack of shadow was useful (thanks), I think adding a border-top to match the border on the room header should suffice, eg

composer_alt-01-reply@2x


**Design changes**

I realised the border radius could be more rounded to sit better with the voice messages component that appear next to them.

- [x] Border radius for input 12px ~~8px~~
- [ ] Background for composer $info-plinth-bg-color
- [ ] Heights and margins and paddings of composer seem different to Figma (when I inspect element mx_MessageComposer mx_GroupLayout mx_MessageComposer_e2eStatus is 77px tall but in Figma its 56px).  I'm not 100% confident I have the perfect values (would only realise after shipping and trying a lot/feedback) but if possible, would be good to match the file for the first effort.  
- [x] Change hint text if possible to `Send message` from `Send a message...`and `Send encrypted message` from `Send an encrypted message...` (the extras seem to be  unnecessary content.
- [ ] If possible, would like to add feedback to the encrypted thing (background changes, colour changes) but understand if it feels a bit random/out of scope of issue 

https://user-images.githubusercontent.com/1587166/172166230-c0d70ee4-d0d9-43b9-8f1a-72c82d31f760.mov

**Comment by gsouquet** [comment]
Sounds good! Will implement those changes.
I would be in favour to deal with the changes to the shield in a different issue. Trying to cross the finish line on this to be able to experiment as early as possible.

Maybe you can open another issue to explain the problem you're trying to solve and we'll pick that up afterwards

**Review by niquewoodhouse — CHANGES_REQUESTED** [review]
- [ ] .mx_MessageComposer_row padding: 8px 12px ~~12px 8px~~
- [ ] .mx_ReplyTile 1.2rem ~~1.4rem~~ 

Is there an easy way to not have this white border at the top when replying?  the border should be the divider between the timeline and the composer

image


I'm seeing the composer encryption badge look misaligned in safari, it looks like its 1px lower than the action buttons. this is really minor and happy to go with it as is.

image

**Pull request merged**